pkg/statedb2/txn.go

```diff
@@ -0,0 +1,369 @@
+}
+
+func (txn *txn) getTable(name TableName) *tableEntry {
+	table, ok := txn.tables[name]
+	if ok {
+		return table
+	}
+	if t, ok := txn.rootReadTxn.Get([]byte(name)); ok {
+		return &t
+	}
+	panic(fmt.Sprintf("BUG: table %q not found", name))
+}
+
+func (txn *txn) addDeleteTracker(meta TableMeta, trackerName string, dt deleteTracker) error {
+	if txn.rootReadTxn == nil {
+		return fmt.Errorf("transaction is closed")
+	}
+	table, ok := txn.tables[meta.Name()]
+	if !ok {
+		return fmt.Errorf("table %q not locked for writing", meta.Name())
+	}
+	dt.setRevision(table.revision)
+	table.deleteTrackers, _, _ = table.deleteTrackers.Insert([]byte(trackerName), dt)
+	return nil
+
+}
+
+func (txn *txn) Delete(meta TableMeta, data any) (any, bool, error) {
+	if txn.rootReadTxn == nil {
+		return nil, false, fmt.Errorf("transaction is closed")
+	}
+
+	tableName := meta.Name()
+	revision, err := txn.newRevision(tableName)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// Delete from the primary index first to grab the object.
+	// We assume that "data" has only enough defined fields to
+	// compute the primary key.
+	idKey := meta.primaryIndexer().fromObject(object{data: data}).First()
+	idIndexTree := txn.indexWriteTxn(tableName, meta.primaryIndexer().name)
+	obj, existed := idIndexTree.Delete(idKey)
+	if !existed {
+		return nil, false, fmt.Errorf("object not found")
+	}
+
+	// Update revision index.
+	indexTree := txn.indexWriteTxn(tableName, RevisionIndex)
+	if _, ok := indexTree.Delete(revisionKey(obj.revision, idKey)); !ok {
+		panic("BUG: Revision entry not found")
+	}
+
+	// Then update secondary indexes.
+	for index, indexer := range meta.secondaryIndexers() {
+		indexer.fromObject(obj).Foreach(func(key []byte) {
+			if !indexer.unique {
+				key = append(key, idKey...)
+			}
+			txn.indexWriteTxn(tableName, index).Delete(key)
+		})
+	}
+
+	// And finally insert the object into the graveyard.
+	if txn.hasDeleteTrackers(tableName) {
+		graveyardIndex := txn.indexWriteTxn(tableName, GraveyardIndex)
+		obj.revision = revision
+		if old, existed := graveyardIndex.Insert(idKey, obj); existed {
+			txn.indexWriteTxn(tableName, GraveyardRevisionIndex).Delete(revisionKey(old.revision, idKey))
+		}
+		txn.indexWriteTxn(tableName, GraveyardRevisionIndex).Insert(revisionKey(revision, idKey), obj)
+	}
+
+	return obj.data, true, nil
+}
+
+func (txn *txn) Abort() {
+	if txn.writeTxns == nil {
+		return
+	}
+	txn.smus.Unlock()
+	*txn = zeroTxn
+}
+
+func (txn *txn) Commit() {
+	if txn.writeTxns == nil {
+		return
+	}
+
+	// Commit each individual changed index to each table.
+	// We don't notify yet (CommitOnly) as the root needs to be updated
+	// first.
+	for tableIndex, subTxn := range txn.writeTxns {
+		table, ok := txn.tables[tableIndex.table]
+		if !ok {
+			panic("BUG: Table " + tableIndex.table + " not cached")
+		}
+		table.indexes, _, _ =
+			table.indexes.Insert([]byte(tableIndex.index), subTxn.CommitOnly())
```

Comment:
you're explicitly ignoring the error here. I think it's non-obvious why this can't fail, so maybe an explanatory comment would help? :)

Reply:
What error? Insert cannot fail.